### ISSUE TITLE
Documentation for handling errors

### ISSUE THREAD

**Issue by valerii15298** [keycloak/keycloak#16900]
### Description

Now it is very difficult to handle errors properly. As I understood when error occurs, axios object is thrown or something like that.
Would be nice to have some error handling functionality or at least docs.

### Motivation

In production apps we need to have reliable way to handle errors and respond correctly.

**Comment by edewit** [comment]
where do you see that @valerii15298 ?

**Comment by valerii15298** [comment]
@edewit here: https://github.com/keycloak/keycloak-nodejs-admin-client/blob/main/package.json#L28

**Comment by edewit** [comment]
@valerii15298 that is the old version the code there has been moved to the main repo here https://github.com/keycloak/keycloak/tree/main/js/libs/keycloak-admin-client

**Comment by valerii15298** [comment]
@edewit thanks!
I suggest you put this link to new code in old repo. So that people will understand where exactly new code is moved.

**Comment by edewit** [comment]
@valerii15298 we already did that already in the readme

**Comment by valerii15298** [comment]
@edewit I am using latest version of `@keycloak/keycloak-admin-client` from npm: `21.0.2`. I checked and it uses `AxiosError` !!!
Why is that so?
Shoud I create new issue for it?

**Comment by valerii15298** [comment]
Latest version on npm uses axios:
https://www.npmjs.com/package/@keycloak/keycloak-admin-client?activeTab=code
as we see from `package.json` and in files it uses axios ....

**Comment by valerii15298** [comment]
Even though this one is closed there is no any solution for this, error handling is not described at all and points described in this issue are not resolved...
As I looked into source code there is `NetworkError` class so I can at least check `e instanceof NetworkError` which is not described in the docs Readme too. The main issue is with the admin rest api itself, there are no docs about error responses(or maybe I missed it?)

And keycloak-admin-client is just a wrapper library around admin rest api...

But in anyway having `unknown_error` and other undocumented responses is far from great dev experience. And absence of error handling documentation makes admin rest api and hence keycloak-admin-client library unreliable.

I suggest to reopen this, or create new one, error handling is not something with low priority and we do not wanna our apps to crash in prod.

Also if there are any ways to contribute for it to accelerate this can someone give me a hint to admin rest api source code and where responses are returned in code? I guess admin rest api is in Java?
If I can check source code maybe I even will be able to make PR for the docs I hope

FYI: @

**Comment by valerii15298** [comment]
FYI: @edewit

**Comment by edewit** [comment]
to be honest the api is not consistent when it comes to errors sometimes the error is a sting sometimes it's a property called message or error. This is what we do to get the error:

```js
if (typeof error === "string") {
    return error;
  }

  if (error instanceof NetworkError) {
    return getNetworkErrorMessage(error);
  }

  if (error instanceof Error) {
    return error.message;
  }

  throw new Error("Unable to determine error message.");
```
```js
function getNetworkErrorMessage({ responseData }) {
  const data = responseData;

  for (const key of ["error_description", "errorMessage", "error"]) {
    const value = data[key];

    if (typeof value === "string") {
      return value;
    }
  }
}
```

**Comment by jonkoops** [comment]
> I am really interested when source code for keycloak admin rest api docs and actual admin api implementation is, so that I can at least explore it and possibly maybe make PRs to improve docs

Not really sure what you are asking for here, this sentence is kinda hard for me to read. But if you are looking to improve the documentation you are always welcome to open up a PR :+1:

**Comment by valerii15298** [comment]
@jonkoops thanks for the response!
I meant where the source code for keycloak admin api implementation located?
And where the source code for the documentation of keycloak admin rest api located?

It seems you already gave me some links to the source code of docs in another thread so I will try to explore it. The codebase is huge so will take quite some time for me. Thanks!!!

**Comment by edewit** [comment]
Code wise everything is here in this repo: 

* [The documentation](https://github.com/keycloak/keycloak/tree/main/docs/documentation)
* [rest admin api](https://github.com/keycloak/keycloak/tree/main/services/src/main/java/org/keycloak/services/resources)
* [admin client](https://github.com/keycloak/keycloak/tree/main/js/libs/keycloak-admin-client)

> Between your approach can lose some info because I encountered two of this keys be present at the same time: ["error_description", "errorMessage", "error"].
I don't know but I think the error_description had similar information as the error, it was only a bit more detailed.

In the end this comes down to having a better admin rest api and we have several issues for that.

**Comment by valerii15298** [comment]
Thanks, @edewit !!!
Can you please assist how the docs for this page https://www.keycloak.org/docs-api/21.1.1/rest-api/index.html created? I guess it is generated from Java code somehow because I cannot find where is it located in docs directory?

**Comment by jonkoops** [comment]
I am re-opening this issue as this is about documenting error handling, which we currently do not have documented.

**Comment by valerii15298** [comment]
@jonkoops I updated issue description to better match the point, since this is the issue about original rest api docs and not the nodejs wrapper library, 😉

**Comment by tylergriffin333** [comment]
I've encountered this same problem.

I've come across some JSON error messages while working with our keycloak instance:

{"errorMessage":"Password policy not met"} 
{"error":"Realm not found."}

but to properly handle all these, I need to know what all to expect.  Are all of these going to be "errorMesage" or "error", or should I expect other message types? 🤷‍♂️

I'm trying to find this information in the keycloak codebase, but there's no one place that these messages come from so far as I can tell.

These sometimes come back with 200 or 201 http response codes, so checking the http status code in the response is not always adequate.

**Comment by jonkoops** [comment]
Yes, these are unfortunately very inconsistent in the API. From what we know there are three possible fields that could return an error string namely `error_description`, `error` and `errorMessage`. You can see we also have to handle this [in our own code](https://github.com/keycloak/keycloak/blob/acfd9305a2393d432865371f257de4757ab24876/integration/client-cli/admin-cli/src/main/java/org/keycloak/client/admin/cli/util/HeadersBodyStatus.java#L56-L58).

**Comment by jonkoops** [comment]
Also, if someone is willing to do so I'd gladly accept a PR that replaces all instances of `error_description` and `errorMessage` with `error`.

**Comment by LouisLau-art** [comment]
Follow-up: this documents NetworkError handling (status + responseData) and the catchNotFound option for 404 -> null behavior in the admin-client README via PR #46822.
